packages/integration-sdk-runtime/src/execution/uploader.test.ts

```diff
@@ -0,0 +1,177 @@
+import { FlushedGraphObjectData } from '../storage/types';
+import {
+  createPersisterApiStepGraphObjectDataUploader,
+  createQueuedStepGraphObjectDataUploader,
+  StepGraphObjectDataUploader,
+} from './uploader';
+import {
+  sleep,
+  createTestEntity,
+  createTestRelationship,
+} from '@jupiterone/integration-sdk-private-test-utils';
+import times from 'lodash/times';
+import { v4 as uuid } from 'uuid';
+import { SynchronizationJobContext } from '../synchronization';
+import { createApiClient, getApiBaseUrl } from '../api';
+import { generateSynchronizationJob } from '../synchronization/__tests__/util/generateSynchronizationJob';
+
+function createFlushedGraphObjectData(): FlushedGraphObjectData {
+  return {
+    entities: [createTestEntity(), createTestEntity()],
+    relationships: [createTestRelationship(), createTestRelationship()],
+  };
+}
+
+async function createAndEnqueueUploads(
+  uploader: StepGraphObjectDataUploader,
+  n: number,
+) {
+  const flushed = times(n, createFlushedGraphObjectData);
+
+  for (const data of flushed) {
+    await uploader.enqueue(data);
+  }
+
+  return flushed;
+}
+
+async function createFlushedDataAndWaitForUploads(
+  uploader: StepGraphObjectDataUploader,
+  n: number,
+) {
+  const flushed = await createAndEnqueueUploads(uploader, n);
+  await uploader.waitUntilUploadsComplete();
+  return flushed;
+}
+
+describe('#createQueuedStepGraphObjectDataUploader', () => {
+  test('should wait for all enqueued uploads to complete', async () => {
+    const uploaded: FlushedGraphObjectData[] = [];
+    let numQueued = 0;
+
+    const uploader = createQueuedStepGraphObjectDataUploader({
+      stepId: uuid(),
+      uploadConcurrency: Infinity,
+      async upload(d) {
+        if (numQueued) {
+          await sleep(100);
+          uploaded.push(d);
+        } else {
+          numQueued++;
+          await sleep(200);
```

Comment:
What's the purpose of these `sleep()`s? It's not very clear to me and makes the test a bit more confusing.

Reply:
Good question. Will add a comment for clarity. The sleeps validate that the promises are indeed executed concurrently instead of in serial. We enqueue a bunch of uploads and, as long as we haven't reached our concurrency limit, the promises will all be run concurrently.